session.go

```diff
@@ -28,81 +29,30 @@ func NewSession(useragent string) *Session {
 }
 
 // DefaultFrontpage returns the submissions on the default reddit frontpage.
-func (s Session) DefaultFrontpage() ([]*Submission, error) {
-	req := request{
-		url:       "http://www.reddit.com/.json",
-		useragent: s.useragent,
-	}
-	body, err := req.getResponse()
-	if err != nil {
-		return nil, err
-	}
-
-	type Response struct {
-		Data struct {
-			Children []struct {
-				Data *Submission
-			}
-		}
-	}
-	r := &Response{}
-	err = json.NewDecoder(body).Decode(r)
-	if err != nil {
-		return nil, err
-	}
-
-	submissions := make([]*Submission, len(r.Data.Children))
-	for i, child := range r.Data.Children {
-		submissions[i] = child.Data
-	}
-
-	return submissions, nil
+func (s Session) DefaultFrontpage(sort popularitySort, params ListingOptions) ([]*Submission, error) {
+	return s.SubredditSubmissions("", sort, params)
 }
 
-// SortedSubmissions return the submission for the request subreddit based on the
-// provided sorting and listing options
-func (s Session) SortedSubmissions(subreddit string, sort popularitySort, params ListingOptions) ([]*Submission, error) {
+// SubredditSubmissions returns the submissions on the given subreddit.
+func (s Session) SubredditSubmissions(subreddit string, sort popularitySort, params ListingOptions) ([]*Submission, error) {
 	v, err := query.Values(params)
 	if err != nil {
 		return nil, err
 	}
-	baseUrl, err := url.Parse(fmt.Sprintf("http://www.reddit.com/r/%s/%s.json?%s", subreddit, sort, v.Encode()))
 
-	req := request{
-		url:       baseUrl.String(),
-		useragent: s.useragent,
-	}
-	body, err := req.getResponse()
-	if err != nil {
-		return nil, err
-	}
+	redditUrl := []string{"http://www.reddit.com", "/%s/%s.json?%s"}
 
-	type Response struct {
-		Data struct {
-			Children []struct {
-				Data *Submission
-			}
-		}
+	if subreddit != "" {
+		/* if a subreddit was given, insert without memory garbage */
+		redditUrl = append(redditUrl, "")
+		copy(redditUrl[2:], redditUrl[1:])
+		redditUrl[1] = "/r"
 	}
```

Comment:
IMO, this whole strategy of building a slice and then using `strings.Join` feels too hard to read.
It'd probably be better to be more verbose, simply to improve readability.